File: problems/problemset/url-router/Python/runner/testbench.py
import sys
import json
import time
import resource

sys.path.insert(0, '/workspace/code')
from router import Router

def main():
    testcase = json.loads(sys.stdin.read())
    router = Router()
    output_lines = []

    start = time.perf_counter()

    for line in testcase['input'].splitlines():
        parts = line.split()
        if not parts:
            continue
        cmd = parts[0]
        if cmd == 'REGISTER':
            router.register(parts[1], parts[2])
        elif cmd == 'MATCH':
            result = router.match(parts[1], parts[2])
            if result is None or not result.pattern:
                output_lines.append('NOTFOUND')
            else:
                out = parts[1] + ' ' + result.pattern
                for k, v in result.params:
                    out += ' ' + k + '=' + v
                output_lines.append(out)

    elapsed_ms = (time.perf_counter() - start) * 1000
    memory_kb = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    actual_output = '\n'.join(output_lines) + '\n' if output_lines else ''

    result = {'actual_output': actual_output, 'time_ms': elapsed_ms, 'memory_kb': memory_kb}
    sys.stdout.write(f'__BEGIN_RESULT__{json.dumps(result)}__END_RESULT__')

main()

File: problems/problemset/url-router/Rust/runner/testbench.rs
mod router;
use router::Router;

use std::io::{self, Read};
use std::time::Instant;

fn extract_json_string(json: &str, key: &str) -> String {
    let needle = format!("\"{}\":", key);
    let start = json.find(&needle).expect("key not found") + needle.len();
    let rest = json[start..].trim_start();
    assert!(rest.starts_with('"'));
    let mut out = String::new();
    let mut chars = rest[1..].chars();
    loop {
        match chars.next().expect("unterminated string") {
            '"' => break,
            '\\' => match chars.next().expect("unterminated escape") {
                'n' => out.push('\n'),
                't' => out.push('\t'),
                'r' => out.push('\r'),
                '"' => out.push('"'),
                '\\' => out.push('\\'),
                c => out.push(c),
            },
            c => out.push(c),
        }
    }
    out
}

fn peak_memory_kb() -> u64 {
    std::fs::read_to_string("/proc/self/status")
        .unwrap_or_default()
        .lines()
        .find(|l| l.starts_with("VmRSS:"))
        .and_then(|l| l.split_whitespace().nth(1))
        .and_then(|s| s.parse().ok())
        .unwrap_or(0)
}

fn escape_json(s: &str) -> String {
    s.replace('\\', "\\\\").replace('"', "\\\"").replace('\n', "\\n").replace('\r', "\\r")
}

fn main() {
    let mut raw = String::new();
    io::stdin().read_to_string(&mut raw).unwrap();

    let test_input = extract_json_string(&raw, "input");
    let mut router = Router::new();
    let mut output_lines: Vec<String> = Vec::new();

    let start = Instant::now();

    for line in test_input.lines() {
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.is_empty() { continue; }
        match parts[0] {
            "REGISTER" => router.register(parts[1], parts[2]),
            "MATCH" => {
                let result = router.match_route(parts[1], parts[2]);
                if result.pattern.is_empty() {
                    output_lines.push("NOTFOUND".into());
                } else {
                    let mut out = format!("{} {}", parts[1], result.pattern);
                    for (k, v) in &result.params {
                        out.push(' ');
                        out.push_str(k);
                        out.push('=');
                        out.push_str(v);
                    }
                    output_lines.push(out);
                }
            }
            _ => {}
        }
    }

    let elapsed_ms = start.elapsed().as_secs_f64() * 1000.0;
    let memory_kb = peak_memory_kb();
    let actual_output = if output_lines.is_empty() { String::new() } else { output_lines.join("\n") + "\n" };

    print!(
        "__BEGIN_RESULT__{{\"actual_output\":\"{}\",\"time_ms\":{:.3},\"memory_kb\":{}}}__END_RESULT__",
        escape_json(&actual_output), elapsed_ms, memory_kb
    );
}
